vdr/ambassador.go

```diff
@@ -37,20 +49,245 @@ type Ambassador interface {
+// This method will check the integrity of the DID document related to the public key used to sign the network document.
+// The rules are based on the Nuts RFC006
+// payload should be a json encoded did.document
+func (n *ambassador) callback(document dag.SubscriberDocument, payload []byte) error {
+	logging.Log().Info("Processing DID documents received from Nuts Network.", DIDDocumentType, document.Ref())
+	if err := checkSubscriberDocumentIntegrity(document); err != nil {
+		return fmt.Errorf("callback could not process new DID Document: %w", err)
+	}
+
+	// Unmarshal the next/new proposed version of the DID Document
+	var nextDIDDocument did.Document
+	if err := json.Unmarshal(payload, &nextDIDDocument); err != nil {
+		return fmt.Errorf("unable to unmarshall did document from network payload: %w", err)
+	}
+
+	if isUpdate(document) {
+		return n.updateDIDDocument(document, nextDIDDocument)
+	}
+	return n.createDIDDocument(document, nextDIDDocument)
+}
+
+func (n *ambassador) createDIDDocument(document dag.SubscriberDocument, nextDIDDocument did.Document) error {
+	// First, check if the network document was signed by the same key as is embedded in the DID Document`s authenticationMethod:
+
+	// Take signing key from network document header
+	headerKey := document.SigningKey()
+	// Create key thumbprint
+	headerKeyThumbprint, err := headerKey.Thumbprint(thumbprintAlg)
+	if err != nil {
+		return fmt.Errorf("unable to generate network document signing key thumbprint: %w", err)
+	}
+
+	// Check if headerKey is one of the keys embedded in the authenticationMethod
+	didDocumentAuthKeys := nextDIDDocument.Authentication
+	if documentKey, err := n.findKeyByThumbprint(headerKeyThumbprint, didDocumentAuthKeys); documentKey == nil || err != nil {
+		if documentKey == nil {
+			return fmt.Errorf("key used to sign Network document must be be part of DID Document authentication")
+		}
+		return err
+	}
+
+	documentMetadata := types.DocumentMetadata{
+		Created:    document.SigningTime(),
+		Updated:    nil,
+		Version:    NewDocumentVersion,
+		TimelineID: document.Ref(),
+		Hash:       document.PayloadHash(),
+	}
+	return n.didStore.Write(nextDIDDocument, documentMetadata)
+}
+
+func (n *ambassador) updateDIDDocument(document dag.SubscriberDocument, nextDIDDocument did.Document) error {
+	// Resolve current version of DID Document
+	resolverMetadata := &types.ResolveMetadata{
+		AllowDeactivated: false,
+	}
+	currentDIDDocument, currentDIDMeta, err := n.didStore.Resolve(nextDIDDocument.ID, resolverMetadata)
+	if err != nil {
+		return fmt.Errorf("unable to update did document: %s", err)
+	}
+	// Check if the new document is actual newer by comparing timeline versions
+	if currentDIDMeta.Version >= document.TimelineVersion() {
+		return fmt.Errorf("unable to update did document: timeline version of current document is greater or equal to the new version")
+	}
+	if currentDIDMeta.TimelineID != document.TimelineID() {
```

Comment:
`TimelineID` is a `SHA256Hash`? Not sure how the `!=` then works, but it has an `Equals` func